app/assets/javascripts/discourse/controllers/preferences/second-factor-backup.js.es6

```diff
@@ -0,0 +1,104 @@
+import { default as computed } from "ember-addons/ember-computed-decorators";
+import { default as DiscourseURL, userPath } from "discourse/lib/url";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+
+export default Ember.Controller.extend({
+  loading: false,
+  errorMessage: null,
+  successMessage: null,
+  backupEnabled: Ember.computed.alias("model.second_factor_backup_enabled"),
+  backupCodes: null,
+
+  @computed("secondFactorToken")
+  isValidSecondFactorToken(secondFactorToken) {
+    return secondFactorToken && secondFactorToken.length === 6;
+  },
+
+  @computed("isValidSecondFactorToken", "backupEnabled", "loading")
+  isDisabledGenerateBackupCodeBtn(isValid, backupEnabled, loading) {
+    return !isValid || loading;
+  },
+
+  @computed("isValidSecondFactorToken", "backupEnabled", "loading")
+  isDisabledDisableBackupCodeBtn(isValid, backupEnabled, loading) {
+    return !isValid || !backupEnabled || loading;
+  },
+
+  @computed("backupEnabled")
+  generateBackupCodeBtnLabel(backupEnabled) {
+    return backupEnabled
+      ? "user.second_factor_backup.regenerate"
+      : "user.second_factor_backup.enable";
+  },
+
+  actions: {
+    copyBackupCode(successful) {
+      if (successful) {
+        this.set(
+          "successMessage",
+          I18n.t("user.second_factor_backup.copied_to_clipboard")
+        );
+      } else {
+        this.set(
+          "errorMessage",
+          I18n.t("user.second_factor_backup.copy_to_clipboard_error")
+        );
+      }
+
+      setTimeout(() => {
```

Comment:
I think you should add a comment here explaining why there's a timeout as I was a little confused at first. It seems to be to hide the message 2 seconds after it's shown?

Also the code should use `Ember.run.later` instead of`setTimeout` is not run within Ember's runloop.